File: em-april-2026/dsa/leetcode_practice_problems/unsorted/03618_split_array_by_prime_indices/main.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

// template <typename T>
// ostream& operator<<(ostream& o, const vector<T>& v) {
// o << "[";
// for (const auto& e : v) {
// o << e << ",";
// }
// return o << "]";
// }

class Solution {
   public:
    long long splitArray(vector<int> nums) {
        auto n = nums.size();
        // cout << << "n = " << n << endl;

        // // profiling
        // int num_additions = 0;
        // int num_lookups = 0;
        // int num_assignments = 0;
        // int num_comparisons = 0;
        // int num_multiplications = 0;

        if (n < 2) {
            return abs(nums[0]);
        }

        // // profiling
        // ++num_lookups;
        // ++num_lookups;
        // ++num_additions;

        long long compositeSum = nums[0] + nums[1];
        // cout << "Added nums[0] and nums[1] (" << nums[0] << " and " <<
        // nums[1]
        // << " ) to compositeSum, so compositeSum = " << compositeSum
        // << endl;
        long long primeSum = 0;
        auto nextPrime = 2;

        // // profiling
        // ++num_comparisons;

        while (nextPrime < n) {
            // // profiling
            // ++num_lookups;
            // ++num_additions;

            primeSum += nums[nextPrime];

            // cout << "Added nums[" << nextPrime << "] ( " << nums[nextPrime]
            // << " ) to primeSum, so primeSum = " << primeSum << endl;

            // // profiling;
            // ++num_assignments;

            nums[nextPrime] = 0;

            // // profiling;
            // ++num_multiplications;

            auto nextCompositeFactor = 2;
            auto nextComposite = nextPrime * nextCompositeFactor;

            // // profiling;
            // ++num_comparisons;

            while (nextComposite < n) {
                // // profiling;
                // ++num_comparisons;
                // ++num_lookups;
                // ++num_additions;

                compositeSum += nums[nextComposite];

                // cout << "Added nums[" << nextComposite << "] ("
                // << nums[nextComposite]
                // << " ) to compositeSum, so compositeSum = " << compositeSum
                // << endl;

                // // profiling;
                // ++num_assignments;

                nums[nextComposite] = 0;

                // // profiling;
                // ++num_multiplications;
                // ++num_additions;

                ++nextCompositeFactor;
                nextComposite = nextPrime * nextCompositeFactor;
            }

            ++nextPrime;

            // // profiling;
            // ++num_lookups;
            // num_comparisons+=2;

            while (nextPrime < n && nums[nextPrime] == 0) {
                // // profiling;
                // num_comparisons+=2;
                // ++num_lookups;

                ++nextPrime;
            }
        }

        // cout << "n = " << n << ", num_additions = " << num_additions
        // << ", num_lookups = " << num_lookups
        // << ", num_multiplications = " << num_multiplications
        // << ",  num_assignments = " << num_assignments
        // << ", num_comparisons = " << num_comparisons << endl;

        return abs(primeSum - compositeSum);
    }
};

int main() {
    // vector<int> nums(100000000, 0);
    // int i = 0;
    // for (auto &elem : nums) {
    // ++elem;
    // ++i;
    // }
    vector<int> nums = {-595161400, -843023220, -292831332,
                        566390112,  945233513,  33600495,
                        865537855,  929580473,  706560008};
    // cout << "nums = " << nums << endl;
    // nums = [2, 3, 4];
    // nums = [-1, 5, 7, 0];
    auto s = Solution();
    auto solution = s.splitArray(nums);
    long long expected_ans =
        abs(nums[0] + nums[1] + nums[4] + nums[6] + nums[8] - nums[2] -
            nums[3] - nums[5] - nums[7]);
    // cout << "solution = " << solution << ", expected answer = " <<
    // expected_ans
    // << endl;
    return 0;
}

File: em-april-2026/dsa/leetcode_practice_problems/unsorted/03618_split_array_by_prime_indices/main.py
class Solution:
    def splitArray(self, nums: list[int]) -> int:
        if not nums:
            return 0
        n = len(nums)
        # print(f"n = {n}")

        # # profiling
        # num_additions = 0
        # num_lookups = 0
        # num_assignments = 0
        # num_comparisons = 0
        # num_multiplications = 0

        if n < 2:
            return abs(nums[0])

        # # profiling
        # num_lookups = num_lookups + 1
        # num_lookups = num_lookups + 1
        # num_additions = num_additions + 1

        compositeSum = nums[0] + nums[1]
        primeSum = 0
        nextPrime = 2

        # # profiling
        # num_comparisons = num_comparisons + 1

        while nextPrime < n:

            # # profiling
            # num_lookups = num_lookups + 1
            # num_additions = num_additions + 1

            primeSum += nums[nextPrime]

            # # profiling
            # num_assignments = num_assignments + 1

            nums[nextPrime] = 0

            # # profiling
            # num_multiplications = num_multiplications + 1

            nextCompositeFactor = 2
            nextComposite = nextPrime * nextCompositeFactor

            # # profiling
            # num_comparisons = num_comparisons + 1

            if nextComposite < n:

                # # profiling
                # num_comparisons = num_comparisons + 1
                # num_lookups = num_lookups + 1
                # num_additions = num_additions + 1

                compositeSum += nums[nextComposite]

                # # profiling
                # num_assignments = num_assignments + 1

                nums[nextComposite] = 0

                # # profiling
                # num_multiplications = num_multiplications + 1
                # num_additions = num_additions + 1

                nextCompositeFactor = nextCompositeFactor + 1
                nextComposite = nextPrime * nextCompositeFactor

            nextPrime = nextPrime + 1

            # # profiling
            # num_lookups = num_lookups + 1
            # num_comparisons = num_comparisons + 2

            while nextPrime < n and nums[nextPrime] == 0:

                # # profiling
                # num_comparisons = num_comparisons + 2
                # num_lookups = num_lookups + 1

                nextPrime = nextPrime + 1

        # print(
        # f"n = {n}, num_additions = {num_additions}, num_lookups = {num_lookups}, num_multiplications = {num_multiplications}, num_assignments = {num_assignments}, num_comparisons = {num_comparisons}"
        # )

        return abs(primeSum - compositeSum)


def main():
    nums = [(i + 1) for i in range(0, 100000)]
    # print(f"nums = {nums}")
    # nums = [2, 3, 4]
    # nums = [-1, 5, 7, 0]
    s = Solution()
    solution = s.splitArray(nums)
    print(f"solution = {solution}")


if __name__ == "__main__":
    main()
